### PR TITLE
Problem: no rapid setup of performant relay nodes

### PR DESCRIPTION
**Pull request by faddat**
# Solution

This PR adds scripts/statesync.bash, which will have a relay-optoimized state synced node running in seconds.  The goal of this commit is to allow relayers to run with the setup that I find ideal:

* State Synced
* RocksDB
* Latest Tendermint

In on order for rocksdb to work correctly, it replaces tecbot/gorocksdb with cosmos/gorocksdb.

It bumps the Tendermint version because .15 seems to do a bit better than .14.

# PR Checklist:

- [x] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [x] Have you rebased your work on top of the latest master? 
- [x] Have you checked your code compiles? (`make`)
- [x] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [x] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

Yes, I agree with contributing.md and will shortly get a little more info on DCO.

Yes, I'm the author of this code :). 

Yes, I like outlining PR's as problem and solution.  Have a nice day!

### CODE REVIEW

**Review by tomtau — COMMENTED** [review]
thanks for the contribution! If I understand it correctly, the problem this is trying to solve is that the default build setup isn't ideal for high performance sentry nodes, is that correct? Could you perhaps open an issue that would have a more detailed problem description and link it here?

Once that's clearer, one thing to consider: why RocksDB and not BadgerDB? (BadgerDB should be supported too and it should be easier to build, given it's pure Go.)

Plus, if the problem is about the default build config, the solution for an alternative build script may be missing e.g. ldflags ( `-X github.com/cosmos/cosmos-sdk/types.DBBackend=...`): https://github.com/cosmos/cosmos-sdk/blob/52f6fc29da96339787da2048bb326b747a71c798/Makefile#L64

**Comment by faddat** [comment]
@tomtau 

Basically, relay nodes across Cosmos -- Osmosis, Gaia, Terra and Cro in particular-- are hit with an unusual isssue.  Nodes used for relaying resemble sentries except that most sentries do not have very much inbound query traffic / tx writes.  Relay nodes have lots of both.

When the IBC network heats up, the node software is no longer able to sync and keep up to the tip of the chain. 

I imagine that you noticed that as the CRO chain gets longer, it gets less and less fast, while consuming more and more disk i/o bandwith, right?

Well, state sync is a way to remedy this issue because you've got only the tip of the chain.  Additionally, state sync makes it possible to -- no kidding -- validate gaia from a raspberry pi.  That's how much disk i/o consumption is reduced (I think by "shortening" the iavl tree, so there's less balancing to do.  So, when you've got a state synced node, another benefit is that it does far better in terms of query performance and writes.  

Typically, I'm in full support of 100% go solutions, but the reason for the ibc traffic jams is basically that relayers like us at Notional or the CROsNest team aren't able to keep nodes synced to the tip of the chain.  So far, my testing has shown that badgerdb is just a little bit faster than goleveldb for use on tendermint chains, while RocksDB is **much** faster, so this particular script is actually part of a devops deployment pattern we're building out at Notional to make our relay setup reproducible ("just add seed phrase") and much more performant than it currently is.  

Long-run, the team at valnodes has proposed:

https://twitter.com/valnodes/status/1466903567013449732?s=20

And I'm in full support of that effort but I figure it is at least a year out currently.

**Review by tomtau — CHANGES_REQUESTED** [review]
ok, could you add the missing parts for the rocksdb build: https://github.com/cosmos/cosmos-sdk/blob/52f6fc29da96339787da2048bb326b747a71c798/Makefile#L76 in the script?

File: scripts/statesync.bash [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+# export state sync vars
+export CHAIN_MAIND_GRPC_WEB_ADDRESS=127.0.0.1:9876
+export CHAIN_MAIND_STATESYNC_ENABLE=true
+export CHAIN_MAIND_P2P_MAX_NUM_OUTBOUND_PEERS=200
+export CHAIN_MAIND_STATESYNC_RPC_SERVERS="https://mainnet.crypto.org:443,https://mainnet.crypto.org:443"
+export CHAIN_MAIND_STATESYNC_TRUST_HEIGHT=$BLOCK_HEIGHT
+export CHAIN_MAIND_STATESYNC_TRUST_HASH=$TRUST_HASH
+export CHAIN_MAIND_P2P_SEEDS="8dc1863d1d23cf9ad7cbea215c19bcbe8bf39702@p2p.baaa7e56-cc71-4ae4-b4b3-c6a9d4a9596a.cryptodotorg.bison.run:26656,8a7922f3fb3fb4cfe8cb57281b9d159ca7fd29c6@p2p.aef59b2a-d77e-4922-817a-d1eea614aef4.cryptodotorg.bison.run:26656,494d860a2869b90c458b07d4da890539272785c9@p2p.fabc23d9-e0a1-4ced-8cd7-eb3efd6d9ef3.cryptodotorg.bison.run:26656,dc2540dabadb8302da988c95a3c872191061aed2@p2p.7d1b53c0-b86b-44c8-8c02-e3b0e88a4bf7.cryptodotorg.herd.run:26656,"
+export CHAIN_MAIND_P2P_PERSISTENT_PEERS="87c3adb7d8f649c51eebe0d3335d8f9e28c362f2@seed-0.crypto.org:26656,e1d7ff02b78044795371beb1cd5fb803f9389256@seed-1.crypto.org:26656"
```
Comment (yihuang) [inline scripts/statesync.bash:39]: nice, I forgot we can override configs in this way.

**Comment by faddat** [comment]
Happy to add the rocksdb config to the Makefile :).

**Comment by tomtau** [comment]
there's a small conflict with the latest master -- @faddat could you rebase your PR after adding `-X github.com/cosmos/cosmos-sdk/types.DBBackend=rocksdb` or what else is needed in the build script?

File: go.mod [hunk 2]
```diff
@@ -145,4 +145,6 @@ replace github.com/cosmos/cosmos-sdk => github.com/cosmos/cosmos-sdk v0.44.5
 
 replace github.com/tendermint/tendermint => github.com/tendermint/tendermint v0.34.15
 
+replace github.com/tecbot/gorocksdb => github.com/cosmos/gorocksdb v0.0.0-20211202124722-2c356d6d98e4
```
Comment (yihuang) [inline go.mod:148]: https://github.com/cosmos/cosmos-sdk/blob/master/db/go.mod#L36
cosmos-sdk use `v1.1.1`, should we follow suite here?

File: scripts/statesync.bash [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+# microtick and bitcanna contributed significantly here.
+set -uxe
+
+# set environment variables
+export GOPATH=~/go
+export PATH=$PATH:~/go/bin
+
+
+# Install chain-maind
+go install -tags rocksdb ./...
```
Comment (yihuang) [inline scripts/statesync.bash:11]: One need to install the rocksdb c library by himself in advance, right?

Comment (yihuang) [reply scripts/statesync.bash:11]: Do we also need the ldflat: `-X github.com/cosmos/cosmos-sdk/types.DBBackend=rocksdb`?

**Comment by tomtau** [comment]
Closing this PR for now -- feel free to reopen it with the extra changes in Makefile

**Pull request closed without merging**